internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,393 @@
+	// but non system errors should be handled internally.
+	// These error semantics allow for proper run time tracking.
+	WorkFunc func(InputType) (OutputType, error) `json:"-"`
+
+	// An optional function to get the locking key for an individual work item.
+	// Provide an input key generation function to synchronize at the input item level in addition to the job level.
+	WorkItemKeyFunc func(InputType) string `json:"-"`
+
+	// An optional function to process the final JobOutput upon successful completion.
+	OnSuccess func(JobOutput[InputType, OutputType]) `json:"-"`
+}
+
+// The output from running a job.
+// A critical error (JobOutput.Error) may not generate work errors, so both must be checked independently.
+// Consult the map of errors using the item to check for item-level errors.
+// Always wait for the Done channel to be closed before handling the output,
+// which can be achieved by calling JobOutput.IsDone().
+type JobOutput[InputType comparable, OutputType any] struct {
+	// Signals the job was canceled during execution.
+	Canceled bool
+
+	// A system error from running the job.
+	Error error
+
+	// A map of errors encountered while running the job.
+	// The key of the error will be the item that failed.
+	WorkErrors map[InputType]error
+
+	// The map of results.
+	ResultItems map[InputType]OutputType
+
+	// The list of items that still need work.
+	RemainingItems []InputType
+
+	// The total computation time spent working on items.
+	// This time does not include time spent waiting for locks or retrieving stored items.
+	// The run time does not include time spent working on items that returned an error.
+	RunTime int64
+
+	// Signals the job is complete.
+	Done <-chan any `json:"-"`
+}
+
+// Use this method to block until the results can be accessed.
+func (this JobOutput[InputType, OutputType]) IsDone() {
+	<-this.Done
+}
+
+func (this *Job[InputType, OutputType]) Validate() error {
+	if this == nil {
+		return fmt.Errorf("Job is nil.")
+	}
+
+	if this.WorkFunc == nil {
+		return fmt.Errorf("Job cannot have a nil work function.")
+	}
+
+	if this.PoolSize <= 0 {
+		return fmt.Errorf("Pool size must be positive, got %d.", this.PoolSize)
+	}
+
+	return this.JobOptions.Validate()
+}
+
+func (this *JobOptions) Validate() error {
+	if this == nil {
+		return fmt.Errorf("Job options are nil.")
+	}
+
+	if this.Context == nil {
+		this.Context = context.Background()
+	}
+
+	return nil
+}
+
+// Process input items in a parallel pool of workers and return the collected results.
+// When not waiting for completion, the output may still be modified until the Done channel is closed.
+// Returning a copy of results gives immediate access to stable results but the final results will be inaccessible.
+// If the context is canceled during execution, the output will signal accordingly.
+func (this *Job[InputType, OutputType]) Run() *JobOutput[InputType, OutputType] {
+	// Run closes the channel to signal the returned output is safe to handle.
```

Comment:
s/Run/Run()/

Suffixing function names with parens makes it easier to identify them.